### DISCUSSION TITLE
What exactly should I do for writing helper functions? (Using Appium + WebDriverIO to test my Android App)

### DISCUSSION THREAD

**Discussion by kainbryanjones** [webdriverio/webdriverio#12048] [Q&A]
I wanted to write some helper functions to abstract repeated tasks (as programmers do), but I kept getting this error...

"My Spec file is not able to load ,getting error as Unable to load spec files quite likely because they rely on browser object that is not fully initialised."

When I moved my function definitions inside the test spec and didn't import them seperately then this worked.

My understanding is that doing this...
`import {driver} from "@wdio/globals"`
outside of a test spec will cause issues.

So what are my options here? Must I pass the driver instance to every helper method I write? Is there an alternative way to reference the driver without encountering this error?

**Comment by christian-bromann** [comment]
@kainbryanjones you can import the browser any time at the top of the file, but you can only use it within a hook or test block. Mocha reads the test file synchronously and doesn't allow for asynchronous executions.

**Reply by kainbryanjones** [reply]
@christian-bromann Thanks for the help!

Perhaps we've discovered some unknown bug then? Basically we (my team) have been writing tests to test our Android app using WebDriverIO + Appium and whenever we've imported the driver instance outside of a *.spec.ts file we are getting errors about using the browser context and invalid imports. I will recreate the issues and post the exact errors below this comment.

For now we've resorted to doing this inside the helper functions.

**const driverToUse = _wdioGlobals.get("driver");**

this isn't ideal, but does work

**Reply by kainbryanjones** [reply]
Common errors:
Unable to load spec files quite likely because they rely on `browser` object that is not fully initialized.
`browser` object has only `capabilities` and some flags like `isMobile`.
Helper files that use other `browser` commands have to be moved to `before` hook.

and
Error: Error: ERR_UNSUPPORTED_DIR_IMPORT

**Reply by christian-bromann** [reply]
@kainbryanjones can you provide a minimal reproducible example?
